feat: multicast udp socket support

jorgehermo9 opened the pull request:
Closes #5732

This PR is still in draft. I have a few pending `TODOs` and also missing tests to propertly check this. Although, this happy path is working.

In order to test this, use this vector config:

```toml
[sources.multicast_udp]
type = "socket"
mode = "udp"
address = "0.0.0.0:4242"
multicast_groups = ["224.0.0.2"]


[sinks.console]
type = "console"
inputs = ["multicast_udp"]
encoding.codec = "json"
```

and with this command
```sh
echo "hello" | nc 224.0.0.2 4242 -u
```

you can see logs in vector.
<img width="494" alt="image" src="https://github.com/user-attachments/assets/764c3224-be79-45a6-a184-64668262755a" />


I would like to receive some feedback about how the configuration of this settings should look like.

Also, note that IPv6 is not supported. We can work on that, but maybe it is not worth it if no one request that.

@nomalord take a look into this please, It would be great if you can build the binary from this branch and test if it works in your systems.

@dalesample as you were the first requester of this I also ping you, just in case (although this issue was created 4 years ago)

jorgehermo9 commented on lib/vector-config/src/stdlib.rs:
```diff
@@ -402,6 +402,31 @@ impl ToValue for SocketAddr {
     }
 }
 
+impl Configurable for Ipv4Addr {
```
Inspired from the `SocketAddr` implementation a few lines above

nomalord commented:
This looks awesome🤯
As it's new years I won't be able to check our usecase until Thursday...
I will say though, that this looks like it answers our usecase exactly🙏🙏🙏🙏🙏🙏

jorgehermo9 commented:
Hi @nomalord perhaps you could test this?

nomalord commented:
I don't really have experience with rust, but I'll try to compile the feature branch and update you🥹

pront commented:
> I don't really have experience with rust, but I'll try to compile the feature branch and update you🥹

Maybe this is helpful to you:
```
# cd /path/to/vector/repo
# brew install gh 
gh pr checkout 22099
cargo run -- --config /path/to/your/config.yaml
```

jorgehermo9 commented:
Thank you for popping in @pront 😃. I have left to include some tests to validate this PR and also add documentation & changelog. I will update this as soon as I can and mark the pr as non-draft then

jorgehermo9 commented on lib/vector-config/src/stdlib.rs:
```diff
@@ -402,6 +402,31 @@ impl ToValue for SocketAddr {
+    }
+
+    fn generate_schema(_: &RefCell<SchemaGenerator>) -> Result<SchemaObject, GenerateError> {
+        // TODO: We don't need anything other than a string schema to (de)serialize a `Ipv4Addr`,
```
Added the same TODOs as in https://github.com/vectordotdev/vector/blob/e63e2bfc1c3f4f86b62274d375236867170bbc6b/lib/vector-config/src/stdlib.rs#L391

pront replied:
Do you plan to add validation in this PR? I understand  we have similar TODOs for `SocketAddr`. If you don't, this is fine with me and we can resolve this PR thread.

jorgehermo9 replied:
I don't, I just copied the `TODO`s a few lines above, but not planned to address it in this PR

jorgehermo9 commented on src/sources/socket/mod.rs:
```diff
@@ -899,12 +899,27 @@ mod test {
+        send_lines_udp_from(next_addr(), addr, lines)
+    }
+
+    fn send_lines_udp_from(
```
refactored this a little bit, because the socket from where we will send the packets must be on the same interface as where the source is listening. And, with multicast groups, `127.0.0.1` seems to not work propertly and we have to use both `0.0.0.0` for the source's listening socket and for the socket that will send the packets. Hope I'm explaining this well..

jorgehermo9 commented on src/sources/socket/udp.rs:
```diff
@@ -152,6 +166,41 @@ pub(super) fn udp(
+                    .join_multicast_v4(group_addr, *listen_addr.ip())
+                    .map_err(|error| {
+                        // TODO: is this considered a `SocketBindError`? or should we create a new error for this case?
+                        emit!(SocketBindError {
```
is this considered a `SocketBindError`? or should we create a new error for this case?

pront replied:
Strictly speaking joining and binding are different. If it's not too much effort, we can introduce a new error and include the group addr.

jorgehermo9 replied:
Agree! Addressed it in dc09eb57fca6e637375c788d2f1517e0cd541f05

jorgehermo9 commented on src/sources/socket/udp.rs:
```diff
@@ -152,6 +166,41 @@ pub(super) fn udp(
+                    // We could support Ipv6 multicast with the
+                    // https://doc.rust-lang.org/std/net/struct.UdpSocket.html#method.join_multicast_v6 method
+                    // and specifying the interface index as `0`, in order to bind all interfaces.
+                    panic!("IPv6 multicast is not supported")
```
I thought about adding ipv6 support too... But as no one requested it yet, I thought i wouldn't be worth to implement & test it in this PR. Ayn thoughts about this? Should I include ipv6 support in this PR?

pront replied:
Good call. Also, a separate PR is better because it make review easier. We can leave this as a potential improvement until there's demand for it. 

```suggestion
                    unimplemented!("IPv6 multicast is not supported")
```

jorgehermo9 replied:
Totally agree!

jorgehermo9 commented on src/sources/socket/udp.rs:
```diff
@@ -152,6 +166,41 @@ pub(super) fn udp(
+                // socket is bound to, and not `IP_ADDR_ANY`. We need to use the same address
+                // for the multicast group join that the user has set in the config.
+                // if systemd{N} fd sockets are required to work too, we should investigate on this.
+                SocketListenAddr::SystemdFd(_) => {
```
We could use the `UdpSocket::local_addr` with this socket, but I wonder if this would ever output `IPADDR_ANY` (0.0.0.0) instead of the real interface address that socket is bound to.

multicast ip required setting 0.0.0.0 as a bind address to work. Otherwise, the interface would filter out packets targeting the multicast ip address. 

I think it would be difficult to get this working as I would have to investigate more about this,  and no one requested it... Any thoughts?

pront replied:
We can leave this as `unimplemented` and potential future work.

jorgehermo9 commented on src/sources/socket/udp.rs:
```diff
@@ -152,6 +166,41 @@ pub(super) fn udp(
+            };
+            for group_addr in config.multicast_groups {
+                socket
+                    .join_multicast_v4(group_addr, *listen_addr.ip())
```
Should I use `listen_addr.ip()` here o just hardcode `Ipv4Addr::UNSPECIFIED`? If `0.0.0.0` is not set, the interface would filter out the packets targeting the multicast ip

see the method documentation here: https://doc.rust-lang.org/std/net/struct.UdpSocket.html#method.join_multicast_v4

and more info about ip multicast here: https://stackoverflow.com/questions/10692956/what-does-it-mean-to-bind-a-multicast-udp-socket

jorgehermo9 replied:
Anyway, if either `listen_addr.ip()` or `Ipv4Addr::UNSPECIFIED` is used here, this would only if the source's address in config is `0.0.0.0` as setting it to `127.0.0.1` (for example) would filter out any packet going to the multicast addres


for example, this config:
```toml
[sources.multicast_udp]
type = "socket"
mode = "udp"
address = "127.0.0.1:4242"
multicast_groups = ["224.0.0.2"]


[sinks.console]
type = "console"
inputs = ["multicast_udp"]
encoding.codec = "json"
```

and this command `echo hello | nc 224.0.0.2 4242 -u`

wont work because the socket address is no `0..0.0.0`. 

If we hardcode this to `Ipv4Addr::UNSPECIFIED`, the behaviour would be the same; that config won't work.

Is this clear enough?

jorgehermo9 commented on src/sources/socket/mod.rs:
```diff
@@ -1303,6 +1318,117 @@ mod test {
+    }
+
+    #[tokio::test]
+    #[should_panic]
```
Had to use `#[should_panic]` here as `init_udp_with_config` does not return a `Result` to know whether the initialization failed or not https://github.com/vectordotdev/vector/blob/12bf074927b960593f7037faaa6ec454dd324214/src/sources/socket/mod.rs#L1018

pront replied:
Makes sense 👍 

TIOLI: Leave a TODO comment to refactor `init_udp_with_config` to return a `Result`

jorgehermo9 commented on src/sources/socket/udp.rs:
```diff
@@ -152,6 +166,33 @@ pub(super) fn udp(
+                            error,
+                        })
+                    })?;
+                info!(message = "Joined multicast group.", group = %group_addr);
```
To give a bit of feedback about group joining

pront replied:
I think it's ok to use `info!` here 

cc @jszwedko to double check from a UX perspective 🙏

jorgehermo9 commented on changelog.d/5732_multicast_udp_socket_sources.feature.md:
```diff
@@ -0,0 +1,8 @@
+The `socket` source with `udp` mode now supports joining multicast groups via the `multicast_groups` option
+of that source. This allows the source to receive multicast packets from the specified multicast groups.
+
+Note that in order to work properly, the `socket` address must be set to `0.0.0.0` and not
```
Is this notice needed if it was added too to website docs?

pront replied:
Thanks for adding this. We can definitely keep it here.

jorgehermo9 commented on src/test_util/mod.rs:
```diff
@@ -120,6 +120,10 @@ pub fn next_addr() -> SocketAddr {
     next_addr_for_ip(IpAddr::V4(Ipv4Addr::LOCALHOST))
 }
 
+pub fn next_addr_any() -> SocketAddr {
```
I had to add this method  due to https://github.com/vectordotdev/vector/pull/22099/files#r1921563985

I needed an address that pointed to `0.0.0.0` instead of `127.0.0.1` as `next_addr()` does.

I thought about changing the `next_addr` implementation to return `Ipv4Addr::UNSPECIFIED` instead of `Ipv4Addr::LOCALHOST`, but a bunch of other test failed and maybe it is not worth to change that much.

pront replied:
good call

jorgehermo9 commented:
Hi @pront, this PR is now ready to review, just notifying, no hurries!

pront commented:
> Hi @pront, this PR is now ready to review, just notifying, no hurries!

Thank you 🚀 I will get to this soon.

pront commented on src/sources/socket/mod.rs:
```diff
@@ -899,12 +899,27 @@ mod test {
+
+    fn send_lines_udp_from(
+        from: SocketAddr,
+        addr: SocketAddr,
```
Nit: 
```suggestion
        to: SocketAddr,
```

Here and bellow

jorgehermo9 replied:
Nice catch!

pront commented on src/sources/socket/mod.rs:
```diff
@@ -1303,6 +1318,114 @@ mod test {
+            init_udp_with_config(tx, config).await;
+
+            // We must send packets to the same interface the `socket_address` is bound to
+            // in order to receive the multicast packets this `from` socket sends
```
Nit:
```suggestion
            // in order to receive the multicast packets the `from` socket sends.
```

pront commented on src/sources/socket/udp.rs:
```diff
@@ -41,6 +43,21 @@ pub struct UdpConfig {
     #[configurable(derived)]
     address: SocketListenAddr,
 
+    /// List of IPv4 multicast groups to join on socket's binding process.
```
Great description 👍

jorgehermo9 replied:
Thank you very much!!!

pront reviewed (approved):
Hey @jorgehermo9, great work. Nothing major came out, mostly nits.

jorgehermo9 commented:
Thanks for the review @pront! I will happily address the comments as I prefer to fix those nits before merging!

jorgehermo9 commented on src/sources/socket/mod.rs:
```diff
@@ -1303,6 +1321,113 @@ mod test {
+
+    #[tokio::test]
+    async fn udp_invalid_multicast_group() {
+        assert_source_error(&COMPONENT_ERROR_TAGS, async {
```
refactored this a little in dc09eb57fca6e637375c788d2f1517e0cd541f05. Didn't know that `assert_source_error` existed. Is it ok to check that only the tag `error_type` is updated? (`COMPONENT_ERROR_TAGS` only contains that tag)

we are using a bit more tags in that error 
![image](https://github.com/user-attachments/assets/c46e5c62-ee3a-4408-9af9-8b617f767eb0)

pront replied:
Yes, for this test it's fine

jorgehermo9 commented on src/internal_events/socket.rs:
```diff
@@ -135,6 +137,37 @@ impl<E: std::fmt::Display> InternalEvent for SocketBindError<E> {
+    pub interface: Ipv4Addr,
+}
+
+impl<E: std::fmt::Display> InternalEvent for SocketMulticastGroupJoinError<E> {
```
Inspired from the `SocketBindError` implementation

jorgehermo9 commented on src/internal_events/socket.rs:
```diff
@@ -135,6 +137,37 @@ impl<E: std::fmt::Display> InternalEvent for SocketBindError<E> {
 }
 
+#[derive(Debug)]
+pub struct SocketMulticastGroupJoinError<E> {
```
I have doubts about this naming. `SocketMulticastJoinError`, `UdpSocketMulticastJoinError`, `UdpSocketMulticastGroupJoinError`....?

This error will be created only from UDP mode, as the multicast concept only exists UDP.

pront replied:
It's a bit verbose but accurate. If you feel strongly towards renaming, feel free to change it.

jorgehermo9 replied:
Agree...Leaving it as-is then! thanks

jorgehermo9 commented on src/internal_events/socket.rs:
```diff
@@ -135,6 +137,44 @@ impl<E: std::fmt::Display> InternalEvent for SocketBindError<E> {
+            "error_type" => error_type::IO_FAILED,
+            "stage" => error_stage::RECEIVING,
+            "mode" => mode,
+            "group_addr" => group_addr,
+            "interface" => interface,
```
Is it ok to add those two variables to counter tags? Or should only be included in the `error!` log?

pront replied:
Answered above

jorgehermo9 commented on src/internal_events/socket.rs:
```diff
@@ -135,6 +137,44 @@ impl<E: std::fmt::Display> InternalEvent for SocketBindError<E> {
+#[derive(Debug)]
+pub struct SocketMulticastGroupJoinError<E> {
+    pub error: E,
+    pub group_addr: Ipv4Addr,
```
Added the `group_addr` and `interface` because they're the arguments required for `join_multicast_v4` 
https://github.com/vectordotdev/vector/blob/a2930d262a76bbe532d53fc356ce442a5a789499/src/sources/socket/udp.rs#L191

pront replied:
These have small low cardinality right? If yes, feel free to resolve this thread.

jorgehermo9 replied:
Yes!

jorgehermo9 commented on src/internal_events/socket.rs:
```diff
@@ -135,6 +137,44 @@ impl<E: std::fmt::Display> InternalEvent for SocketBindError<E> {
+impl<E: std::fmt::Display> InternalEvent for SocketMulticastGroupJoinError<E> {
+    fn emit(self) {
+        // Multicast groups are only used in UDP mode
+        let mode = SocketMode::Udp.as_str();
```
Is it ok to hardcode Udp here? This error can only happen in udp protocol

jorgehermo9 commented on src/internal_events/socket.rs:
```diff
@@ -135,6 +137,44 @@ impl<E: std::fmt::Display> InternalEvent for SocketBindError<E> {
+            error = %self.error,
+            error_code = "socket_multicast_group_join",
+            error_type = error_type::IO_FAILED,
+            stage = error_stage::RECEIVING,
```
error_stage receiving? I found that stage weird in this case. I think `INITIALIZING` or so would fit more in this case... but I see that `error_stage::RECEIVING` is used also for socket binding error and etc.. Just noticed this and wanted to raise it, not for addressing in this PR though

pront replied:
`INITIALIZING` is slightly more clear for the end user IMO

jorgehermo9 replied:
okay, changed it for the binding error too. Done in c03bcfb9a

jorgehermo9 commented:
Hi @pront. Sorry for the delay, had no time these past weeks. Addressed all your comments and left a few doubts about the new changes. Feel free to resolve any conversation, just wanted to raise some things so you are aware of it.

Thanks!

jorgehermo9 commented:
small bump in case this got forgotten @pront, I think it is almost done.

jorgehermo9 commented:
Ready to go! Thanks @pront

CC @nomalord I hope you still need this, sorry for the delay!

jorgehermo9 commented:
Hm, It is surprising that `multicast_and_unicast_udp_message` fails but `multicast_udp_message` doesn't fail. Both uses `224.0.0.2` multicast address, so I don't think that address is the problem. The only difference is that `multicast_and_unicast_udp_message` sends an unicast packet after the multicast one here: 
https://github.com/vectordotdev/vector/blob/cdd37cdf977353281d604f0722150f4c62437f3f/src/sources/socket/mod.rs#L1409

But can't understand why it is not working if `multicast_udp_message` is. I can only think of `next_addr_any` returning an invalid address... But it makes no sense :/

Is there any easy way to run windows tests from linux? Or do I have to spin up a VM and configure everything there..?

What we could do is to flag that test behind #[cfg(unix)] or delete it completely, as I think that test is a bit redundant, as we have `multicast_udp_message` and `udp_message` tests though.

jorgehermo9 commented:
Can't understand why the error is happening, as we always bind to `INADDR_ANY` and not to an specific IP [see google search for this error](https://www.google.com/search?q=The+requested+address+is+not+valid+in+its+context.+%28os+error+10049%29+0.0.0.0&sca_esv=bb3cec438e37b9d9&sxsrf=AHTn8zoWiEwmNZnZXbbez0wWPUg4c2VWGw%3A1741128085998&ei=lYHHZ-HNPN_c7_UPqpeImA4&ved=0ahUKEwih7MvIv_GLAxVf7rsIHaoLAuMQ4dUDCBA&uact=5&oq=The+requested+address+is+not+valid+in+its+context.+%28os+error+10049%29+0.0.0.0&gs_lp=Egxnd3Mtd2l6LXNlcnAiS1RoZSByZXF1ZXN0ZWQgYWRkcmVzcyBpcyBub3QgdmFsaWQgaW4gaXRzIGNvbnRleHQuIChvcyBlcnJvciAxMDA0OSkgMC4wLjAuMEjyGVC7BVj2F3ADeACQAQGYAWOgAY4MqgECMjC4AQPIAQD4AQGYAgWgArMBwgIKEAAYsAMY1gQYR8ICBBAjGCfCAgQQABgemAMAiAYBkAYFkgcBNaAHjUE&sclient=gws-wiz-serp)

pront commented:
> Or do I have to spin up a VM and configure everything there..?

I am afraid this is the way to go.
 
> What we could do is to flag that test behind #[cfg(unix)] or delete it completely, as I think that test is a bit redundant, as we have `multicast_udp_message` and `udp_message` tests though.

I wonder if this feature works on Windows, sounds like we never tested it on there. I have a feeling ti does but we don't want to merge it without testing it. If you don't want to go down that rabbit hole, we can start by advertising it as linux only (and apply `#[cfg(unix)]` everywhere).

jorgehermo9 commented:
no problem, I will try to run it all on windows!

jorgehermo9 commented:
Hi @pront, spent a few hours trying to configure my windows machine to run the tests but didn't manage to get it working. I also tried to run the tests in github OSS runners, but they lack of enough memory and had OOM while compiling. 

I'm not sure how I could test this, and I dont have enough time to get this compiled on windows. Do you want to close the PR, or what should we do with this?

pront commented:
> Hi @pront, spent a few hours trying to configure my windows machine to run the tests but didn't manage to get it working. I also tried to run the tests in github OSS runners, but they lack of enough memory and had OOM while compiling.
> 
> I'm not sure how I could test this, and I dont have enough time to get this compiled on windows. Do you want to close the PR, or what should we do with this?

Thank you for your efforts here! 
We can make it unix with `#[cfg(unix)]` since it is tested there and we have higher confidence that it will be functional there.

jorgehermo9 commented:
@pront After long hours of debugging, it seems that windows cannot use `0.0.0.0` as an address to connect!

> 0.0.0.0 is not a valid IP address. [RFC1700](https://www.rfc-editor.org/rfc/rfc1700) (a) states that 0.0.0.0/8 (0.anything.anything.anything) is reserved as a source address only.

(https://superuser.com/a/1368920)

But that worked on linux...

This was fixed in https://github.com/vectordotdev/vector/commit/9d7b2af0e0247469bc2d0d1c3116386e788640d3

Previously the test did:

from `0.0.0.0:{fromPort}` -> `224.0.0.2:{targetPort}
from `0.0.0.0:{fromPort}` -> `0.0.0.0:{targetPort` (this was the listening port, `socket_address` variable, and the connection to `0.0.0.0` failed

Now, the test does:

from `0.0.0.0:{fromPort}` -> `224.0.0.2:{targetPort}
from `0.0.0.0:{fromPort}` -> `127.0.0.1:{targetPort` (we can connect to 127.0.0.1 on windows)

It is okay that the `from` part binds to `0.0.0.0`, the problem is _connecting to_ `0.0.0.0`.....

However, I'm glad the problem was just the test and not the implementation... Sorry for this being delayed so much, I had some problems building on windows (spent a few hours to build from sourcebefore noticing that I had to build with `--no-default-features --features msvc` due to jemalloc....) and then doing the debugging part...


Thank you so much!
